code/modules/projectiles/guns/projectile/constructable/blastcannon.dm

```diff
@@ -0,0 +1,225 @@
+	icon = 'icons/obj/gun.dmi'
+	icon_state = "blastcannon_empty"
+	inhand_states = list("left_hand" = 'icons/mob/in-hand/left/guninhands_left.dmi', "right_hand" = 'icons/mob/in-hand/right/guninhands_right.dmi')
+	item_state = null
+	w_class = 3.0
+	force = 5
+	flags = FPRINT
+	siemens_coefficient = 1
+	ejectshell = 0
+	caliber = null
+	ammo_type = null
+	fire_sound = null
+	conventional_firearm = 0
+	var/bomb = null
+	var/datum/gas_mixture/bomb_air_contents_1 = null
+	var/datum/gas_mixture/bomb_air_contents_2 = null
+	var/ignorecap = 0
+	var/bomb_appearance = null
+
+/obj/item/weapon/gun/projectile/blastcannon/Destroy()
+	if(bomb)
+		qdel(bomb)
+		bomb = null
+	bomb_appearance = null
+	bomb_air_contents_1 = null
+	bomb_air_contents_2 = null
+	..()
+
+/obj/item/weapon/gun/projectile/blastcannon/attack_self(mob/user as mob)
+	if(!bomb)
+		return
+	else
+		var/obj/item/device/transfer_valve/T = bomb
+		T.loc = user.loc
+		user.put_in_hands(T)
+		bomb = null
+		to_chat(user, "You detach \the [T] from \the [src].")
+		bomb_appearance = null
+		name = "pipe gun"
+		desc = "A pipe welded onto a gun stock. You're not sure how you could even use this."
+	update_icon()
+
+/obj/item/weapon/gun/projectile/blastcannon/update_icon()
+	overlays.len = 0
+	if(!bomb || !bomb_appearance)
+		return
+
+	var/image/bomb_icon = image('icons/obj/weaponsmithing.dmi', src, "nothing")
+	bomb_icon.appearance = bomb_appearance
+	bomb_icon.pixel_x = 2
+	bomb_icon.pixel_y = 9
+
+	overlays += bomb_icon
+
+/obj/item/weapon/gun/projectile/blastcannon/attackby(obj/item/weapon/W, mob/user)
+	if(istype(W, /obj/item/device/transfer_valve))
+		var/obj/item/device/transfer_valve/T = W
+		if(!T.tank_one || !T.tank_two)
+			to_chat(user, "<span class='warning'>Nothing's going to happen if there[!T.tank_one && !T.tank_two ? " aren't any tanks" : "'s only one tank"] attached to \the [W]!</span>")
+			return
+		bomb_appearance = W.appearance
+		if(!user.drop_item(W, src))
+			to_chat(user, "<span class='warning'>You can't let go of \the [W]!</span>")
+			bomb_appearance = null
+			return 1
+		bomb = W
+		user.visible_message("[user] attaches \the [W] to \the [src].","You attach \the [W] to \the [src].")
+		name = "blast cannon"
+		desc = "A weapon of devastating force, the explosive power from the tank transfer valve is funneled straight out of its barrel."
+	update_icon()
+
+/obj/item/weapon/gun/projectile/blastcannon/afterattack(atom/A as mob|obj|turf|area, mob/living/user as mob|obj, flag, params, struggle = 0)
+	if (istype(A, /obj/item/weapon/storage/backpack ))
+		return
+
+	else if (A.loc == user.loc)
+		return
+
+	else if (A.loc == user)
+		return
+
+	else if (locate (/obj/structure/table, src.loc))
+		return
+
+	if(!bomb)
+		return
+
+	else
+		var/obj/item/device/transfer_valve/T = bomb
+		if(T.damaged)
+			click_empty(user)
+			return
+
+		if(flag)	return //we're placing gun on a table or in backpack
```

Comment:
Erm, but you already have a check for both a bit higher, and in the case where this would trigger the `click_empty()` would still be called.